include/seastar/core/semaphore.hh

```diff
@@ -165,37 +166,94 @@ private:
+            : sem(&sem_), nr(nr_)
+        {
+            sem->add_waiter(*this);
+        }
+        basic_waiter(basic_semaphore& sem_, size_t nr_, time_point timeout) noexcept
+            : sem(&sem_), nr(nr_)
+        {
+            set_timeout(timeout);
+            sem->add_waiter(*this);
+        }
+        basic_waiter(basic_semaphore& sem_, size_t nr_, abort_source& as) noexcept
+            : sem(&sem_), nr(nr_)
+        {
+            subscribe(as);
+            sem->add_waiter(*this);
+        }
+        virtual ~basic_waiter() = default;
+
+        void init(basic_semaphore& sem_, size_t nr_) noexcept {
+            sem = &sem_;
+            nr = nr_;
+            sem->add_waiter(*this);
+        }
+
+        void subscribe(abort_source& as) noexcept {
+            _aborter.template emplace<optimized_optional<abort_source::subscription>>(as.subscribe([this] (const std::optional<std::exception_ptr>& opt_ex) noexcept {
+                std::exception_ptr ex;
                 try {
-                    e.pr.set_exception(sem.timeout());
+                    if (opt_ex) {
+                        ex = *opt_ex;
+                    } else if (sem) {
+                        if (sem->_state->ex) {
+                            ex = sem->_state->ex;
+                        } else if constexpr (internal::has_aborted<ExceptionFactory>::value) {
+                            ex = std::make_exception_ptr(sem->aborted());
+                        }
+                    }
                 } catch (...) {
-                    e.pr.set_exception(semaphore_timed_out());
                 }
-            } else if (sem._state->ex) {
-                e.pr.set_exception(sem._state->ex);
-            } else {
-                if constexpr (internal::has_aborted<exception_factory>::value) {
-                    try {
-                        e.pr.set_exception(static_cast<exception_factory>(sem).aborted());
-                    } catch (...) {
-                        e.pr.set_exception(semaphore_aborted());
+                abort(ex ? std::move(ex) : std::make_exception_ptr(semaphore_aborted()));
+            }));
+            assert(std::get<optimized_optional<abort_source::subscription>>(_aborter));
+        }
+
+        void set_timeout(time_point timeout) noexcept {
+            if (timeout == time_point::max()) {
+                return;
+            }
+            _aborter.template emplace<timer>([this] () {
+                std::exception_ptr ex;
+                try {
+                    if (sem) {
+                        ex = std::make_exception_ptr(sem->timeout());
                     }
-                } else {
-                    e.pr.set_exception(semaphore_aborted());
+                } catch (...) {
                 }
-            }
+                abort(ex ? std::move(ex) : std::make_exception_ptr(semaphore_timed_out()));
+            });
+            auto& tmr = std::get<timer>(_aborter);
```

Comment:
should probably just use the return value of `_aborter.emplace()` instead of using `std::get<>`, which performs runtime check.